File: python/middle/single_number_ii.py
class Solution(object):
    def singleNumber(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        ret = 0
        if len(nums) == 0:
            return ret
        bit_sum = 0
        for i in xrange(0, 32):
            bit_sum = 0
            for num in nums:
                bit_val = (num>>i) & 1
                bit_sum += bit_val
            bit_sum = bit_sum % 3
            if i == 31 and bit_sum:
                ret -= 1 << 31
            else:
                ret |= bit_sum << i
        return ret
        
        
Note:
In python, an integer is unlimit large, so we can't use the last index(i==31) 0 or 1 to judge whether or not a number is positive or negative.
thus, we need to manually adjust it.

File: cpp/middle/single_number_ii.cpp
class Solution {
public:
    int singleNumber(vector<int>& nums) {
        int ret = 0;
        if(nums.size() == 0)
            return ret;
        int sum = 0;
        for(int i = 0; i < 32; ++i)
        {
            sum = 0;
            for(int j = 0; j < nums.size(); ++j)
            {
                int bit_val = (nums[j] >> i) & 1;
                sum += bit_val;
            }
            sum %= 3;
            ret |= (sum << i);
        }
        return ret;
    }
};


Conclusion:
an middle level algorithm question. Given an input array, all the numbers appear 3 times except only 1 number appear 1 time, our goal is to
find out the one which only appear 1 time in the input array.
In this case, because integer number is 4 bytes=>32 bits; then we can caculate each position (0 ~ 31) how many 1 on each bit, then divide by 3
(because the other number all appear 3 times in the input array.) then we organize each bit number to get the result.

time complexity: O(32N) => O(N)
space complexity: O(1)
